File: src/editor.rs
use std::fs;
use std::path::{Path, PathBuf};

pub struct EditorBuffer {
    pub path: Option<PathBuf>,
    pub lines: Vec<String>,
    pub cursor_x: usize,
    pub cursor_y: usize,
    pub scroll_x: usize,
    pub scroll_y: usize,
    pub modified: bool,
}

impl EditorBuffer {
    pub fn new() -> Self {
        Self {
            path: None,
            lines: vec![String::new()],
            cursor_x: 0,
            cursor_y: 0,
            scroll_x: 0,
            scroll_y: 0,
            modified: false,
        }
    }

    pub fn load<P: AsRef<Path>>(&mut self, path: P) -> Result<(), String> {
        let path_ref = path.as_ref();
        let content = if path_ref.exists() {
            fs::read_to_string(path_ref).map_err(|e| format!("Failed to read file: {}", e))?
        } else {
            String::new()
        };

        let mut lines: Vec<String> = content.lines().map(|s| s.to_string()).collect();
        if lines.is_empty() {
            lines.push(String::new());
        }

        self.path = Some(path_ref.to_path_buf());
        self.lines = lines;
        self.cursor_x = 0;
        self.cursor_y = 0;
        self.scroll_x = 0;
        self.scroll_y = 0;
        self.modified = false;
        Ok(())
    }

    pub fn save(&mut self) -> Result<(), String> {
        if let Some(ref path) = self.path {
            let content = self.lines.join("\n");
            fs::write(path, content).map_err(|e| format!("Failed to write file: {}", e))?;
            self.modified = false;
            Ok(())
        } else {
            Err("No file path set to save".to_string())
        }
    }

    pub fn insert_char(&mut self, c: char) {
        if self.cursor_y >= self.lines.len() {
            self.lines.push(String::new());
        }
        let line = &mut self.lines[self.cursor_y];
        
        // Handle cursor x out of bounds safety
        if self.cursor_x > line.len() {
            self.cursor_x = line.len();
        }
        
        line.insert(self.cursor_x, c);
        self.cursor_x += 1;
        self.modified = true;
    }

    pub fn insert_tab(&mut self) {
        for _ in 0..4 {
            self.insert_char(' ');
        }
    }

    pub fn insert_newline(&mut self) {
        let line = &self.lines[self.cursor_y];
        let (left, right) = line.split_at(self.cursor_x.min(line.len()));
        let left_str = left.to_string();
        let right_str = right.to_string();
        
        self.lines[self.cursor_y] = left_str;
        self.lines.insert(self.cursor_y + 1, right_str);
        self.cursor_y += 1;
        self.cursor_x = 0;
        self.modified = true;
    }

    pub fn backspace(&mut self) {
        if self.cursor_x > 0 {
            let line = &mut self.lines[self.cursor_y];
            if self.cursor_x <= line.len() {
                line.remove(self.cursor_x - 1);
            }
            self.cursor_x -= 1;
            self.modified = true;
        } else if self.cursor_y > 0 {
            let current_line = self.lines.remove(self.cursor_y);
            self.cursor_y -= 1;
            let prev_len = self.lines[self.cursor_y].len();
            self.lines[self.cursor_y].push_str(&current_line);
            self.cursor_x = prev_len;
            self.modified = true;
        }
    }

    pub fn delete(&mut self) {
        let line = &mut self.lines[self.cursor_y];
        if self.cursor_x < line.len() {
            line.remove(self.cursor_x);
            self.modified = true;
        } else if self.cursor_y < self.lines.len() - 1 {
            let next_line = self.lines.remove(self.cursor_y + 1);
            self.lines[self.cursor_y].push_str(&next_line);
            self.modified = true;
        }
    }

    pub fn move_cursor_up(&mut self) {
        if self.cursor_y > 0 {
            self.cursor_y -= 1;
            self.clamp_cursor_x();
        }
    }

    pub fn move_cursor_down(&mut self) {
        if self.cursor_y < self.lines.len() - 1 {
            self.cursor_y += 1;
            self.clamp_cursor_x();
        }
    }

    pub fn move_cursor_left(&mut self) {
        if self.cursor_x > 0 {
            self.cursor_x -= 1;
        } else if self.cursor_y > 0 {
            self.cursor_y -= 1;
            self.cursor_x = self.lines[self.cursor_y].len();
        }
    }

    pub fn move_cursor_right(&mut self) {
        let line = &self.lines[self.cursor_y];
        if self.cursor_x < line.len() {
            self.cursor_x += 1;
        } else if self.cursor_y < self.lines.len() - 1 {
            self.cursor_y += 1;
            self.cursor_x = 0;
        }
    }

    fn clamp_cursor_x(&mut self) {
        let line_len = self.lines[self.cursor_y].len();
        if self.cursor_x > line_len {
            self.cursor_x = line_len;
        }
    }

    pub fn adjust_scroll(&mut self, height: usize, width: usize) {
        if self.cursor_y < self.scroll_y {
            self.scroll_y = self.cursor_y;
        }
        if height > 0 && self.cursor_y >= self.scroll_y + height {
            self.scroll_y = self.cursor_y - height + 1;
        }
        if self.cursor_x < self.scroll_x {
            self.scroll_x = self.cursor_x;
        }
        if width > 0 && self.cursor_x >= self.scroll_x + width {
            self.scroll_x = self.cursor_x - width + 1;
        }
    }
}

File: editor.go
package main

import (
	"os"
	"path/filepath"
	"strings"
	"unicode/utf8"
)

type EditorBuffer struct {
	Path     string
	Lines    []string
	CursorX  int
	CursorY  int
	ScrollX  int
	ScrollY  int
	Modified bool
}

func NewEditorBuffer() EditorBuffer {
	return EditorBuffer{Lines: []string{""}}
}

func (e *EditorBuffer) Load(path string) error {
	content, err := os.ReadFile(path)
	if err != nil {
		if !os.IsNotExist(err) {
			return err
		}
		content = nil
	}

	text := strings.ReplaceAll(string(content), "\r\n", "\n")
	text = strings.TrimSuffix(text, "\n")
	lines := strings.Split(text, "\n")
	if len(lines) == 0 || (len(lines) == 1 && lines[0] == "") {
		lines = []string{""}
	}

	e.Path = path
	e.Lines = lines
	e.CursorX = 0
	e.CursorY = 0
	e.ScrollX = 0
	e.ScrollY = 0
	e.Modified = false
	return nil
}

func (e *EditorBuffer) Save() error {
	if e.Path == "" {
		return errNoFile
	}
	if err := os.MkdirAll(filepath.Dir(e.Path), 0o755); err != nil {
		return err
	}
	if err := os.WriteFile(e.Path, []byte(strings.Join(e.Lines, "\n")), 0o644); err != nil {
		return err
	}
	e.Modified = false
	return nil
}

func (e *EditorBuffer) InsertRune(r rune) {
	e.ensureLine()
	line := []rune(e.Lines[e.CursorY])
	e.CursorX = clamp(e.CursorX, 0, len(line))
	line = append(line[:e.CursorX], append([]rune{r}, line[e.CursorX:]...)...)
	e.Lines[e.CursorY] = string(line)
	e.CursorX++
	e.Modified = true
}

func (e *EditorBuffer) InsertText(s string) {
	for _, r := range s {
		if r == '\n' {
			e.InsertNewline()
			continue
		}
		if r != '\r' {
			e.InsertRune(r)
		}
	}
}

func (e *EditorBuffer) InsertTab() {
	e.InsertText("    ")
}

func (e *EditorBuffer) InsertNewline() {
	e.ensureLine()
	line := []rune(e.Lines[e.CursorY])
	e.CursorX = clamp(e.CursorX, 0, len(line))
	left := string(line[:e.CursorX])
	right := string(line[e.CursorX:])
	e.Lines[e.CursorY] = left
	e.Lines = append(e.Lines[:e.CursorY+1], append([]string{right}, e.Lines[e.CursorY+1:]...)...)
	e.CursorY++
	e.CursorX = 0
	e.Modified = true
}

func (e *EditorBuffer) Backspace() {
	e.ensureLine()
	if e.CursorX > 0 {
		line := []rune(e.Lines[e.CursorY])
		e.CursorX = clamp(e.CursorX, 0, len(line))
		line = append(line[:e.CursorX-1], line[e.CursorX:]...)
		e.Lines[e.CursorY] = string(line)
		e.CursorX--
		e.Modified = true
		return
	}
	if e.CursorY > 0 {
		current := e.Lines[e.CursorY]
		e.Lines = append(e.Lines[:e.CursorY], e.Lines[e.CursorY+1:]...)
		e.CursorY--
		prevLen := utf8.RuneCountInString(e.Lines[e.CursorY])
		e.Lines[e.CursorY] += current
		e.CursorX = prevLen
		e.Modified = true
	}
}

func (e *EditorBuffer) Delete() {
	e.ensureLine()
	line := []rune(e.Lines[e.CursorY])
	if e.CursorX < len(line) {
		line = append(line[:e.CursorX], line[e.CursorX+1:]...)
		e.Lines[e.CursorY] = string(line)
		e.Modified = true
		return
	}
	if e.CursorY < len(e.Lines)-1 {
		e.Lines[e.CursorY] += e.Lines[e.CursorY+1]
		e.Lines = append(e.Lines[:e.CursorY+1], e.Lines[e.CursorY+2:]...)
		e.Modified = true
	}
}

func (e *EditorBuffer) MoveUp() {
	if e.CursorY > 0 {
		e.CursorY--
		e.clampCursorX()
	}
}

func (e *EditorBuffer) MoveDown() {
	if e.CursorY < len(e.Lines)-1 {
		e.CursorY++
		e.clampCursorX()
	}
}

func (e *EditorBuffer) MoveLeft() {
	if e.CursorX > 0 {
		e.CursorX--
		return
	}
	if e.CursorY > 0 {
		e.CursorY--
		e.CursorX = utf8.RuneCountInString(e.Lines[e.CursorY])
	}
}

func (e *EditorBuffer) MoveRight() {
	e.ensureLine()
	if e.CursorX < utf8.RuneCountInString(e.Lines[e.CursorY]) {
		e.CursorX++
		return
	}
	if e.CursorY < len(e.Lines)-1 {
		e.CursorY++
		e.CursorX = 0
	}
}

func (e *EditorBuffer) AdjustScroll(height, width int) {
	if height < 1 {
		height = 1
	}
	if width < 1 {
		width = 1
	}
	if e.CursorY < e.ScrollY {
		e.ScrollY = e.CursorY
	}
	if e.CursorY >= e.ScrollY+height {
		e.ScrollY = e.CursorY - height + 1
	}
	if e.CursorX < e.ScrollX {
		e.ScrollX = e.CursorX
	}
	if e.CursorX >= e.ScrollX+width {
		e.ScrollX = e.CursorX - width + 1
	}
	if e.ScrollY < 0 {
		e.ScrollY = 0
	}
	if e.ScrollX < 0 {
		e.ScrollX = 0
	}
}

func (e *EditorBuffer) clampCursorX() {
	e.ensureLine()
	e.CursorX = min(e.CursorX, utf8.RuneCountInString(e.Lines[e.CursorY]))
}

func (e *EditorBuffer) ensureLine() {
	if len(e.Lines) == 0 {
		e.Lines = []string{""}
	}
	if e.CursorY < 0 {
		e.CursorY = 0
	}
	for e.CursorY >= len(e.Lines) {
		e.Lines = append(e.Lines, "")
	}
}
